NAMES:IRAKOZE Furaha

ID:20251SEN168

PROGRAM:LAB3


Part 1 · List Comprehensions (6 marks)


In [2]:
temps_list = [24.4, 25.9, 28.6, 19.3, 26.0, 22.1]
raw = ["24.4", "25.9", "28.6", "19.3"]

# 1.1 Fahrenheit
temps_f = [t * 9/5 + 32 for t in temps_list]
print(temps_f)          # [75.92, 78.62, 83.48, 66.74, 78.8, 71.78]



[75.92, 78.62, 83.48, 66.74000000000001, 78.8, 71.78]


In [3]:

# 1.2 Filter
hot = [t for t in temps_list if t >= 26]
print(hot)              # [28.6, 26.0]
cool_rounded = [round(t) for t in temps_list if t < 24]
print(cool_rounded)     # [19, 22]



[28.6, 26.0]
[19, 22]


In [4]:


# 1.3 Convert types
floats = [float(x) for x in raw]
print(floats, sum(floats) / len(floats))   # [24.4, 25.9, 28.6, 19.3] 24.55

[24.4, 25.9, 28.6, 19.3] 24.55


Part 2 · lambda 

In [1]:
# 2.1 rank with sorted
results = [("Aline", 85), ("Eric", 78), ("Diane", 91), ("Samuel", 92), ("Grace", 76)]
ranked = sorted(results, key=lambda r: r[1], reverse=True)
print(ranked[:3])   # [('Samuel', 92), ('Diane', 91), ('Aline', 85)]


[('Samuel', 92), ('Diane', 91), ('Aline', 85)]


In [2]:

# 2.2 max with key
rain_log = [(3, 0.0), (7, 12.5), (11, 3.2), (14, 26.2), (18, 8.9)]
wettest = max(rain_log, key=lambda p: p[1])
print(wettest)      # (14, 26.2)

(14, 26.2)


Part 3 · NumPy Arrays 

In [3]:
# 3.1 the speed test
import time
import numpy as np

n = 1_000_000

# Python loop on a plain list
python_list = list(range(n))
start = time.time()
result_loop = []
for x in python_list:
    result_loop.append(x * 2)
loop_time = time.time() - start

# NumPy: one vectorized operation
arr = np.arange(n)
start = time.time()
result_np = arr * 2
numpy_time = time.time() - start

print(f"Python loop: {loop_time:.4f} seconds")
print(f"NumPy:       {numpy_time:.6f} seconds")
print(f"NumPy was about {loop_time / numpy_time:.0f}x faster")

# COMMENT (the part that is graded):
# On my machine NumPy was roughly ___x faster than the Python loop
# because it runs the multiplication in compiled C code on a whole array
# at once, instead of one Python step per number.

Python loop: 0.1379 seconds
NumPy:       0.002436 seconds
NumPy was about 57x faster


In [8]:

# 3.2 describe the temperature
print(f"Mean: {temps.mean():.2f} °C")   # 24.00
print(f"Max:  {temps.max():.1f} °C")    # 28.6
print(f"Min:  {temps.min():.1f} °C")    # 19.3
print(f"Std:  {temps.std():.2f} °C")    # 2.61



Mean: 24.00 °C
Max:  28.6 °C
Min:  19.3 °C
Std:  2.61 °C


In [6]:
import os
print(os.getcwd())
print(os.listdir())


C:\Users\HP\BigDataLab
['.ipynb_checkpoints', 'BigDataCourse', 'Untitled.ipynb', 'venv', 'Week1_IRAKOZEFuraha.ipynb', 'week3_IRAKOZE_Furaha.ipynb', 'week3_report.txt', 'week3_students.csv', 'week5_IRAKOZEFuraha.ipynb', 'week5_kigali_weather (1).csv']


In [7]:
import numpy as np

data = np.genfromtxt("week5_kigali_weather (1).csv", delimiter=",", skip_header=1, dtype=str)
dates = data[:, 1]
temps = data[:, 2].astype(float)
rain  = data[:, 3].astype(float)
hum   = data[:, 4].astype(float)

print(temps[:5])   # should show [24.4 25.9 24.5 26.  24.6]

[24.4 25.9 24.5 26.  24.6]


In [9]:

# 3.3 the hottest day
pos = np.argmax(temps)
print(f"Hottest day: {dates[pos]} at {temps[pos]} °C")   # 2026-09-17 at 28.6 °C

Hottest day: 2026-09-17 at 28.6 °C


Part 4 · Boolean Masking 

Exercise 4.1 — Hot days 


In [10]:
# 4.1 hot days
hot_mask = temps >= 26
print("Hot days:", hot_mask.sum())                          # 18
print("Avg temp on hot days:", temps[hot_mask].mean())       # 27.11


Hot days: 18
Avg temp on hot days: 27.11111111111111


In [11]:

# 4.2 rain report
rainy = rain > 0
print("Rainy days:", rainy.sum())                            # 16
print("Total rainfall:", rain.sum())                         # 331.2 mm
print("Heaviest day:", rain.max())                           # 26.2 mm



Rainy days: 16
Total rainfall: 331.20000000000005
Heaviest day: 26.2


In [12]:
# 4.3 cross-array analysis
print("Avg temp rainy:", temps[rainy].mean())                # 24.14
print("Avg temp dry:  ", temps[~rainy].mean())               # 23.95
print("Hot AND rainy:", ((temps >= 26) & (rain > 0)).sum())  # 5
print("Humidity >= 70:", (hum >= 70).sum())                  # 16

Avg temp rainy: 24.137500000000003
Avg temp dry:   23.94545454545455
Hot AND rainy: 5
Humidity >= 70: 16


Part 5 · Anomalies & Trends 
Exercise 5.1 — Anomaly detection

In [14]:
# 5.1
anomaly = temps - temps.mean()
print(dates[anomaly > 3])
# ['2026-09-07' '2026-09-09' '2026-09-11' '2026-09-12' '2026-09-14'
#  '2026-09-16' '2026-09-17' '2026-09-19' '2026-09-22']   -> 9 days
print("Days > 3°C below mean:", (anomaly < -3).sum())       # 8



['2026-09-07' '2026-09-09' '2026-09-11' '2026-09-12' '2026-09-14'
 '2026-09-16' '2026-09-17' '2026-09-19' '2026-09-22']
Days > 3°C below mean: 8


In [15]:
# 5.2
first, last = temps[:30].mean(), temps[30:].mean()
print(f"First 30 days: {first:.2f} °C")   # 26.21
print(f"Last 30 days:  {last:.2f} °C")    # 21.78
# Kigali got COOLER: the average dropped by about 4.4 °C from September to October.

First 30 days: 26.21 °C
Last 30 days:  21.78 °C


Bonus (+3) · Top 5 wettest days

In [16]:
wettest5 = sorted(zip(dates, rain), key=lambda p: p[1], reverse=True)[:5]
for d, r in wettest5:      # a loop is allowed for printing the bonus
    print(d, r, "mm")

2026-09-07 26.2 mm
2026-09-14 25.1 mm
2026-09-21 24.0 mm
2026-09-05 22.9 mm
2026-09-28 22.9 mm


Part 6 · Reflection

1. What did NumPy hide, and why learn the loop first?
temps.mean() hides the loop, the running total, the counter and the division. NumPy runs those steps in fast compiled code. Learning the loop first means you understand what mean() actually does, so you can debug it, handle odd cases like missing or bad values, and tell when a loop is genuinely needed. Without that, NumPy is a black box.

2. One insight:
Kigali cooled and became more humid over the 60 days. The average temperature fell from 26.2 °C to 21.8 °C, while average humidity rose from about 58% to 66%. Rainfall was frequent in both halves, with 8 rainy days each. Rainy and dry days have almost the same average temperature (24.1 vs 23.9 °C), so the cooling follows the season rather than individual rain events.

Short report (½–1 page, draft)

Insight: Kigali cooled by about 4 °C between September and October 2026.
Over 60 days, mean temperature was 24.0 °C (range 19.3–28.6 °C, std 2.6). The hottest day was 17 September (28.6 °C). Comparing halves, the first 30 days averaged 26.2 °C and the last 30 averaged 21.8 °C. Eighteen days reached 26 °C or more, and all of them fall in the first half of the period. Nine days were more than 3 °C above the mean and eight were more than 3 °C below it, and the warm and cool anomalies cluster at the two ends of the season. It rained on 16 of 60 days (331.2 mm in total), with the heaviest day at 26.2 mm on 7 September. Rain did not make days noticeably cooler (24.1 °C on rainy days vs 23.9 °C on dry days), which suggests the temperature drop is a seasonal trend. Humidity rose from about 58% to 66% as temperatures fell.

(The "all in the first half" point checks out: the last day at 26 °C or above is day 27.)




this is the end